ddl/table.go

```diff
@@ -395,3 +397,104 @@ func updateVersionAndTableInfo(t *meta.Meta, job *model.Job, tblInfo *model.Tabl
 	}
 	return ver, t.UpdateTable(job.SchemaID, tblInfo)
 }
+
+// TODO: It may have the issue when two clients concurrently add partitions to a table.
+func onAddTablePartition(t *meta.Meta, job *model.Job) (ver int64, _ error) {
+	partInfo := &model.PartitionInfo{}
+	err := job.DecodeArgs(&partInfo)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+
+	tblInfo, err := getTableInfo(t, job, job.SchemaID)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+	err = checkPartitionNotExists(tblInfo, partInfo)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+
+	updatePartitionInfo(partInfo, tblInfo)
+	ver, err = updateVersionAndTableInfo(t, job, tblInfo, true)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+	// Finish this job.
+	job.FinishTableJob(model.JobStateDone, model.StatePublic, ver, tblInfo)
+	return ver, errors.Trace(err)
+}
+
+func updatePartitionInfo(partitionInfo *model.PartitionInfo, tblInfo *model.TableInfo) {
+	parInfo := &model.PartitionInfo{}
+	oldDefs, newDefs := tblInfo.Partition.Definitions, partitionInfo.Definitions
+	parInfo.Definitions = make([]model.PartitionDefinition, 0, len(newDefs)+len(oldDefs))
+	parInfo.Definitions = append(parInfo.Definitions, oldDefs...)
+	parInfo.Definitions = append(parInfo.Definitions, newDefs...)
+	tblInfo.Partition.Definitions = parInfo.Definitions
+}
+
+func checkPartitionNotExists(meta *model.TableInfo, part *model.PartitionInfo) error {
+	oldDefs, newDefs := meta.Partition.Definitions, part.Definitions
+	for _, oldDef := range oldDefs {
+		for _, newDef := range newDefs {
+			if strings.EqualFold(oldDef.Name, newDef.Name) {
+				return ErrSameNamePartition.GenByArgs(newDef.Name)
+			}
+		}
+	}
+
+	for i := 0; i < len(newDefs); i++ {
+		for j := i + 1; j < len(newDefs); j++ {
+			if i != j {
+				if strings.EqualFold(newDefs[i].Name, newDefs[j].Name) {
+					return ErrSameNamePartition.GenByArgs(newDefs[i].Name)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// checkAddPartitionValue values less than value must be strictly increasing for each partition.
+func checkAddPartitionValue(meta *model.TableInfo, part *model.PartitionInfo) error {
+	if meta.Partition.Type == model.PartitionTypeRange {
+		newDefs, oldDefs := part.Definitions, meta.Partition.Definitions
+		nextRangeValue := oldDefs[len(oldDefs)-1].LessThan[0]
+
+		if strings.EqualFold(nextRangeValue, "MAXVALUE") {
+			return ErrPartitionMaxvalue
+		}
+
+		if len(newDefs) == 1 && strings.EqualFold(newDefs[0].LessThan[0], "MAXVALUE") {
```

Comment:
Is this `if` redundant?  Below `if ifMaxvalue && i == len(newDefs)-1 ` has the same effect.